### PR TITLE
graphql: support simple arguments like `offset`, `limit` and `distinct_on`

### LINKED ISSUE #12: graphql: support `limit`, `offset` and `distinct_on` clauses in a query
Support queries like this one: 

```graphql

query {
  user(distinct_on: name, limit: 10, offset: 10) {
    name
    uid
  }
}


```

### LINKED ISSUE #23: graphql/json: check casing of values in responses
Functionality-wise:
This might be sth we didn't catch earlier-on:
When I send this payload
```graphql
{
    limitUsers: users(limit: 1) {
        first_name
        last_name
    }
}
```
I get this back with the users' alias changed to lowercase
```json
{
    "data": {
        "limitusers": [
            {
                "first_name": "Ikechukwu",
                "last_name": "Eze"
            }
        ]
    }
}
```

_Originally posted by @iykekings in https://github.com/kolharsam/graphql-engine-rs/pull/22#issuecomment-907883254_

### PR DESCRIPTION
**Pull request by kolharsam**
This PR does some clean-up of the code and resolves #12 and resolves #23 too

To test:
Schema: users(id serial primary key, first_name text, last_name text)

```graphql
{
    limitUsers: users(limit: 2) {
        first_name
        last_name
    }
    distinctUsers: users(distinct_on: last_name) {
        first_name
        last_name
    }
    offsetUsers: users(offset: 3) {
        first_name
        last_name
    }
}
```

### CODE REVIEW

File: src/db.rs [hunk 1]
```diff
@@ -6,40 +6,96 @@ pub fn get_pg_client(connection_string: String) -> Client {
...
+
+            if (*field_arg) == "offset" {
+                let arg_val = field_info.root_field_arguments.get("offset");
+                match arg_val {
+                    None => {}
+                    Some(val) => {
+                        query.push_str(format!("OFFSET {} ", val.get_num()).as_str());
+                    }
+                }
+            }
```
Comment (iykekings) [inline src/db.rs:83]: Consider this for readability and also ease of adding new arguments
```suggestion
           let arg_val = field_info.root_field_arguments.get(*field_arg);
            match *field_arg {
                "limit" => {
                    if let Some(val) = arg_val {
                        query.push_str(format!("LIMIT {} ", val.get_num()).as_str());
                    }
                }
                "offset" => {
                    if let Some(val) = arg_val {
                        query.push_str(format!("OFFSET {} ", val.get_num()).as_str());
                    }
                }
                _ => {}
            }
```

File: src/db.rs [hunk 2]
```diff
@@ -6,40 +6,96 @@ pub fn get_pg_client(connection_string: String) -> Client {
...
+    // ----- Run Query
+
+    let query_result = client.query_one(query.as_str(), &[]);
+
+    match query_result {
+        Ok(res) => Ok(res),
+        Err(err) => Err(error::GQLRSError::new(error::GQLRSErrorType::DBError(
+            format!("{:?}", err),
+        ))),
+    }
```
Comment (iykekings) [inline src/db.rs:100]:
```suggestion
    query_result
        .map_err(|err| error::GQLRSError::new(error::GQLRSErrorType::DBError(format!("{:?}", err))))
```

File: src/server.rs [hunk 3]
```diff
@@ -35,13 +38,16 @@ pub struct ErrorResponse {
...
+    let payload_to_str = std::str::from_utf8(&payload);
+    let parse_result = match payload_to_str {
+        Ok(r) => json::parse(r),
+        Err(_err) => Err(json::Error::FailedUtf8Parsing),
+    };
 
-    // FIXME?: is this even necessary?
     let body: json::JsonValue = match parse_result {
         Ok(v) => v,
         Err(e) => json::object! { "error" => e.to_string() },
```
Comment (iykekings) [inline src/server.rs:53]: Not part of this PR: but we can refactor this:
```suggestion
   let body = parse_result.unwrap_or_else(|e| json::object! { "error" => e.to_string() });
```

Comment (iykekings) [inline src/server.rs:53]: Not part of this PR: but we can refactor this:
```suggestion
   let body = parse_result.unwrap_or_else(|e| json::object! { "error" => e.to_string() });
```

**Review by iykekings — CHANGES_REQUESTED** [review]
Code review completed

**Comment by iykekings** [comment]
Functionality-wise:
This might be sth we didn't catch earlier-on:
When I send this payload
```graphql
{
    limitUsers: users(limit: 1) {
        first_name
        last_name
    }
}
```
I get this back with the users' alias changed to lowercase
```json
{
    "data": {
        "limitusers": [
            {
                "first_name": "Ikechukwu",
                "last_name": "Eze"
            }
        ]
    }
}
```

**Comment by iykekings** [comment]
Everything works as expected (offset, limit and distinct_on). Great work @kolharsam ❤️‍🔥

File: src/types.rs [hunk 4]
```diff
@@ -137,3 +148,99 @@ fn field_name_to_sql_with_alias() {
...
+pub fn to_string_arg<'a>(
+    arg_name: String,
+    arg_val: &graphql_parser::query::Value<'a, &'a str>,
+) -> Result<(String, GQLArgType), error::GQLRSError> {
+    if let graphql_parser::query::Value::Enum(st) = arg_val {
+        return Ok((arg_name, GQLArgType::String(st.to_string())));
+    }
+
+    Err(error::GQLRSError::new(error::GQLRSErrorType::GenericError(
+        "failed to parse string".to_string(),
```
Comment (iykekings) [inline src/types.rs:218]: Better error message
```suggestion
        format!("failed to parse argument {}", arg_name),
```

File: src/db.rs (hunk 1 above)
Comment (kolharsam) [reply src/db.rs:83]: Hey Ike! Thanks for the suggestions! 
I also thought if I should probably create an inline function for this, perhaps? Instead of duplicating the code anywhere at all?

File: src/db.rs (hunk 2 above)
Comment (kolharsam) [reply src/db.rs:100]: This is great! Thanks 😄

File: src/server.rs (hunk 3 above)
Comment (kolharsam) [reply src/server.rs:53]: Since there's some clean up being done in the PR itself, I think it should be fine. I can create a function for this purpose too, since we might be doing this in all other endpoints that consume JSON as request payloads.

**Review by iykekings — APPROVED** [review]
❤️🔥
